### ISSUE TITLE
Instructions for launching Sandboxie Plus, especially portable mode

### ISSUE THREAD

**Issue by Gitoffthelawn** [sandboxie-plus/Sandboxie#62]
Thank you for creating this fork and for all your hard work.

I'm trying to understand the details of how to actually use this tool, especially in portable mode.

After compiling it yourself (or extracting files from the release archive) into a folder, you will have 11 executables.  Which one to use to launch Sandboxie Plus?

Also, to make it portable, is it as simple as creating an empty `Sandboxie-Plus.ini` in Sanboxie Plus's folder?

From looking at the code, it seems like there are 2 UIs provided.  Are they both compatible with the portable mode?

I can update the readme with a summary of answers provided here.

Thanks!

**Comment by DavidXanatos** [comment]
Sandboxie-Plus is started using the SandMan.exe (Sandboxie Manager)

On first start the tool will unpack the driver if its not already unpacked and install it, for the later operation it will request admin privileges.
Once its started and the driver is loaded its ready to go.

To make it behave portable you need to create a Sandboxie-Plus.ini as well as a Sandboxie.ini in the application root folder.

When the tool detects that is running in fully portable mode on exit it will always uninstall the driver.

The legacy UI SbieCtrl.exe can be used once the driver is loaded but it does not have the functionality to Install/Uninstall/Unload) the driver for that SandMan.exe or kmdutil.exe must be used.

I would say if you want to use it portable only use SandMan.exe

**Comment by Gitoffthelawn** [comment]
Thanks so much David!

Before you had time to write your response, I did everything you said except creating Sandboxie.ini in the application root folder.  I did create Sanboxie-Plus.ini.

I'll run through everything to see what changed, but it never hurts to ask: What do you think got changed in the system by not creating Sandboxie.ini first?

**Comment by DavidXanatos** [comment]
A Sandboxie.ini was created under c:\windows you can move it or delete it, that's all.

**Comment by Gitoffthelawn** [comment]
Thank you.  That's all I have found so far.  Almost done checking for anything else.

Would you like me to update the readme for you?

**Comment by DavidXanatos** [comment]
Yes please that would be helpfull

**Comment by Gitoffthelawn** [comment]
I will work on it for you.

Would you also like me to include a link to your Patreon in it?
For anyone reading this, it is:
https://www.patreon.com/DavidXanatos

BTW, I'm a little embarrassed by this, but I was computing without any sleep, and I accidentally ran `kmdutil.exe` from Windows Explorer.  Since no params were provided, it just opened up a message box complaining about the lack of params.  Anything I need to clean up from that?  BTW, what does that executable do?

Also, using a non-sandboxed file manager, I was looking through the sandbox folder to learn more about how Sandboxie is working , and my fatigued fingers, eyes, and brain accidentally launched a shortcut in there that pointed to an executable in a subfolder of the sandbox.  Did I run that shortcut and the executable outside of the sandbox?

*Note to self: Sleep is important. Very important.*

**Comment by DavidXanatos** [comment]
yes please include a link

kmdutil.exe installs/uninstalls and starts/stops the SbieDrv and SbieSvc without parameters it does not do anything except complaining.

When the SbieDrv is loaded starting a exe inside a sandbox location will force it to run sandboxes.

**Comment by Gitoffthelawn** [comment]
> yes please include a link

Will do.

> kmdutil.exe installs/uninstalls and starts/stops the SbieDrv and SbieSvc without parameters it does not do anything except complaining.

Thank you.

> When the SbieDrv is loaded starting a exe inside a sandbox location will force it to run sandboxes.

That's impressive.  So if SandMan is running without error, and you launch Windows Explorer ***outside*** of SandMan, and then you use Windows Explorer (which is not sandboxed)to launch any exe inside the sandbox folder, it will still get run inside the sandbox.  Do I have it correct?

**Comment by DavidXanatos** [comment]
> That's impressive. So if SandMan is running without error, and you launch Windows Explorer _**outside**_ of SandMan, and then you use Windows Explorer (which is not sandboxed)to launch any exe inside the sandbox folder, it will still get run inside the sandbox. Do I have it correct?

Only the driver + service need to be running is the tool is not being executed in portable mode, you don't even have to have SandMan.exe running

Just keep in mind that in fully portable mode sandman on close will stop and remove the driver + service, than it wont work.

**Comment by MCHALMEI** [comment]
Hi, 

I am from Brazil. Thanks for your dev efforts. 

The "SBIE2303 Could not hook CoGetObject (33, 1655)
SBIE2303 Could not hook RegisterDragDrop (33, 1655)
SBIE2318 DLL initialization failed for 'ole32.dll' " is still persisting with Chrome browsers. A pity. 

Suggestion: How about making it possible for users to set a global hotkey combo for cleaning any given Sandbox contents? It's a real pain having to open the program main window for achieving this.

Thanks in advance for your attention. 

Regards.

**Comment by Gitoffthelawn** [comment]
@MCHALMEI Thank you for your report. May I recommend copy and pasting what you wrote to a new issue, and then deleting your text from this issue?

**Comment by Gitoffthelawn** [comment]
@DavidXanatos Thank you for the details.  There were a couple typos in what you wrote, but I think I understand what you are conveying.  Here is what I think you meant to write, in case others have trouble understanding it:

> Only the driver + service need to be running when the tool is not being executed in portable mode.  In non-portable mode, you don't even have to have SandMan.exe running

> Just keep in mind that in fully portable mode, SandMan will stop and remove the driver + service when you close it.  At that point, anything you run will not be sandboxed.

Did I "translate" your words accurately?

**Comment by DavidXanatos** [comment]
yes

**Comment by AndhikaWB** [comment]
> Only the driver + service need to be running is the tool is not being executed in portable mode, you don't even have to have SandMan.exe running

> Just keep in mind that in fully portable mode sandman on close will stop and remove the driver + service, than it wont work.

By executing `KmdUtil.exe` manually to install/uninstall Sandboxie driver (based on [this](https://github.com/sandboxie-plus/Sandboxie/issues/11#issuecomment-615885266) and [this](https://www.hybrid-analysis.com/sample/a7bd6c2218d06034e7ef30d52f28638bd83e2d874f4150fb3827eaa09b23a00e?environmentId=5#signature-96da17c8a0e3b0b5330080ed267d9751)), I believe a restart is actually required on each (un)installation. Sandboxie may not complain about it, but `KmdUtil.exe` does. Is it actually safe to ignore the restart process?

Here is screenshot after uninstalling driver and installing it again (without restarting):
![image](https://user-images.githubusercontent.com/17119394/85229247-d38f0d00-b412-11ea-8fcf-58598bdaeed4.png)

**Comment by DavidXanatos** [comment]
If the sbie driver is unloaded and you are running as administrator no reboot is required.

a reboot needed message usually indicated that the sbie driver wasnt unloaded.

**Comment by Gitoffthelawn** [comment]
> By executing `KmdUtil.exe` manually to install/uninstall Sandboxie driver...

I assume you are using the command-line arguments to do this, correct?

**Comment by DavidXanatos** [comment]
yes you must tell the utillity what to do using command line arguments

**Comment by unconditional** [comment]
@Gitoffthelawn any update on the readme progress?

Starting with just [this info](https://github.com/sandboxie-plus/Sandboxie/issues/62#issuecomment-646498037) will be very useful for everyone, I'm sure.

**Comment by Gitoffthelawn** [comment]
@DavidXanatos Hi David.  Much has transpired since we last communicated in this issue report.  I'm glad to see you are still doing well.

With the recently released v0.5.1, how does one configure Sandboxie Plus to work in portable mode?

With previous versions, users could unpack the executable with any zip tool and copy the needed files.  Then they could create empty `Sandboxie.ini` & `Sandboxie-Plus.ini` files and manually run `SandMan.exe`.

How can portable mode be achieved with v0.5.1?

**Comment by DavidXanatos** [comment]
Probably not that easy as I haven't found a way to get a inno setup to behave as a extract only.
I guess I'll just go back to distributing a zip along side the installer.

**Comment by Gitoffthelawn** [comment]
@DavidXanatos Yes, Inno Setup has its ups and downs.  Lack of easy extractability is probably its biggest downside.  Zip distributions would be great.  Thanks.

**Comment by DavidXanatos** [comment]
With the new pre release build 0.5.3a we have an extractable inno setup yea!
Also some issues around portable usage have been fixed

**Comment by Gitoffthelawn** [comment]
Thank you David!  So just extract the inno setup file like any zip archive and follow the procedures you and I outlined above?

**Comment by DavidXanatos** [comment]
Just run the inno setup and it will offer you to extract without installation to a folder of your choosing
you can add /CURRENTUSER /PORTABLE=1 to the command line to not require admin rights

**Comment by Gitoffthelawn** [comment]
@DavidXanatos Ah... I'm glad I asked.  Thanks.  Add `/CURRENTUSER /PORTABLE=1` to the command line of inno setup or the executable that is extracted by it?

**Comment by DavidXanatos** [comment]
to inno setup than it starts without asking for admin rights

**Comment by geoslake** [comment]
Hi guys, thanks for the portable option !
However it doesnt work here, when i start Sandman.exe, it says "failed to connect to the driver".
This is on a fresh windows 10 install, hidden admin account.
Thanks

**Comment by DavidXanatos** [comment]
just try again, or look in the maintenance menu whats not running booth service and driver should be started, if not try starting the service (it will autostart the driver on its own)

**Comment by geoslake** [comment]
Thanks David. When I try to start the service in the maintenance menu, it says :
ControlService Start: The filename, directory name, or volume lable syntax is incorrect (123)

**Comment by DavidXanatos** [comment]
Try uninstalling and re instilling the service and driver
also verify that the driver path in the registry has \??\ prepended to it.

**Comment by geoslake** [comment]
Mmm i thought it was portable ? Are we talking about regkeys present while sb+ is running ?
Can you give me the regkey location ? i dont find any related to driver path.

Thanks !

**Comment by DavidXanatos** [comment]
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\SbieDrv
ImagePath should start with \??\
if it does nto just prepend it and try starting it again.

**Comment by geoslake** [comment]
"\??\D:\System\Sandboxie+\SbieDrv.sys"

So sounds good ?

**Comment by DavidXanatos** [comment]
```\??\D:\System\Sandboxie+\SbieDrv.sys```

**Comment by geoslake** [comment]
Sorry yes, i miscopied it seems, weird...anyway i confirm that its now
"\??\D:\System\Sandboxie+\SbieDrv.sys"
and still the sane issues

**Comment by DavidXanatos** [comment]
hmm... strange... try rebooting and may be try an other file location although that probably wont help, but in case.

**Comment by Gitoffthelawn** [comment]
@DavidXanatos  David, can we temporarily re-open this issue to improve its visibility?

**Comment by geoslake** [comment]
@DavidXanatos i rebooted many times, tried moving the folder to C:....

**Comment by Gitoffthelawn** [comment]
@geoslake Would it be correct to understand that neither of those things helped?

**Comment by geoslake** [comment]
@Gitoffthelawn yes sorry, thats what i meant. I can try on other pc's today. Im using windows 10 ltsc just in case - as i had another issue specific to that verison it seems https://github.com/sandboxie-plus/Sandboxie/issues/3

**Comment by DavidXanatos** [comment]
@geoslake Running under the build in administrator seams to be problematic, running under any other administrator user even with UAC disabled strangely does not causes issues like in #3

**Comment by geoslake** [comment]
@DavidXanatos Ok so I tried under a standard admin account and it didnt help. I did install the non-plus sandboxie and it failed to run, briging the same error message. Then i uninstalled sandboxie and reinstalled it, now it works. I guess it got rid of erroneous registry entries or something

**Comment by Gitoffthelawn** [comment]
The registry entries part of this is what confuses/concerns me.  What registry entries are being created in the portable version, and is there a way to avoid this?

**Comment by DavidXanatos** [comment]
> 
> 
> The registry entries part of this is what confuses/concerns me. What registry entries are currently being created in the portable version, and is there a way to avoid this so it is fully portable?

HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\SbieDrv
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\SbieSvc

Not these are absolutely unavoidable  the only thing we can do is to delete them once we are done, and sandman.exe should offer to do that when being closed from portable mode.


At some later point I plan to create a driver-less version but this would be only suitable for application virtualization not for security as without the driver we can not guarantee any security, so it wouldn't be a sandbox anymore.

**Comment by Gitoffthelawn** [comment]
Thanks so much David.  For the first word in your first sentence, you meant to type *"Note"*, correct (has the opposite meaning of what is currently posted!)?

**Comment by myfingerhurt** [comment]
I have the same issue.
Confirmed in Windows 10 Sandbox mode.
Sandman works if I have preferred to run it in the installed mode, but Sandman unable to start driver if I have started it with portable mode. 

`Windows 10 2004`
`Version 10.0.19041 Build 19041`

By the way SandMan always complains about 
`SBIE1411 Sandbox DefaultBox specifies unknown template WindowsFontCache`

**Comment by wevsty** [comment]
@myfingerhurt 
> By the way SandMan always complains about
> `SBIE1411 Sandbox DefaultBox specifies unknown template WindowsFontCache`


I have encountered the same problem regarding this issue, which I discuss in #395  .

**Comment by smithy49** [comment]
In the classic Sandboxie I could stop sandboxed Chrome from accessing non-sandboxed chrome data.
This was done under Resource Access -> Write-only access and adding **C:\Users\xxxx\AppData\Local\Google\Chrome\User Data\Default**. 
This stopped the sandboxed Chrome from reading the open tabs in non-sandboxed Chrome.
Is there something similar in Sandoxie-plus?

Thanks in advance for any help.

**Comment by DavidXanatos** [comment]
> 
> 
> @myfingerhurt
> 
> > By the way SandMan always complains about
> > `SBIE1411 Sandbox DefaultBox specifies unknown template WindowsFontCache`
> 
> I have encountered the same problem regarding this issue, which I discuss in #395 .

should be fixed in the latest build

**Comment by DavidXanatos** [comment]
> 
> 
> In the classic Sandboxie I could stop sandboxed Chrome from accessing non-sandboxed chrome data.
> This was done under Resource Access -> Write-only access and adding **C:\Users\xxxx\AppData\Local\Google\Chrome\User Data\Default**.
> This stopped the sandboxed Chrome from reading the open tabs in non-sandboxed Chrome.
> Is there something similar in Sandoxie-plus?
> 
> Thanks in advance for any help.

yes ofcause:

![grafik](https://user-images.githubusercontent.com/3890945/106030884-e2d3e880-60ce-11eb-9a22-d514194a96b3.png)

**Comment by smithy49** [comment]
Hi David,
Thanks for replying.
I tried your example but it didn't work.
Here's what I did.
I pressed the down on Add File/Folder then browsed to folder clicked Select Folder 
then clicked Apply and the line that appeared in Resource Access disappeared.
Maybe this is a bug.
Thanks again.
Charles Smith
    On Wednesday, 27 January 2021, 17:39:15 GMT, DavidXanatos <notifications@github.com> wrote:  
 
 



In the classic Sandboxie I could stop sandboxed Chrome from accessing non-sandboxed chrome data.
This was done under Resource Access -> Write-only access and adding C:\Users\xxxx\AppData\Local\Google\Chrome\User Data\Default.
This stopped the sandboxed Chrome from reading the open tabs in non-sandboxed Chrome.
Is there something similar in Sandoxie-plus?

Thanks in advance for any help.


yes ofcause:

**Comment by DavidXanatos** [comment]
its a semmi bug, you need to press enter or click someware else in the list to save the values to the list than apply saves them to ini.

if you press apply while editing a line the line changes wannt take

**Comment by smithy49** [comment]
Hi David,
I managed to work it out for myself and it did take.
However the isolation did not happen.
I deleted the sandbox contents and cleared all browser cookies and history but 
when I started chrome in the browser it picked up all the open tabs from non-sandboxed chrome.
Thanks for your good work on Sandboxie-Plus.
Charles Smith

    On Thursday, 28 January 2021, 15:09:07 GMT, DavidXanatos <notifications@github.com> wrote:  
 
 


its a semmi bug, you need to press enter or click someware else in the list to save the values to the list than apply saves them to ini.

if you press apply while editing a line the line changes wannt take

**Comment by smithy49** [comment]
Hi David,
When I open the .ini file I see this.

OpenFilePath=C:\Users\name\AppData\Local\Google\Chrome\User Data\Default.

However this is not stopping sandboxed Chrome from accessing the folder above of the non-sandboxed Chrome.

**Comment by Gitoffthelawn** [comment]
> However this is not stopping sandboxed Chrome from accessing the folder above of the non-sandboxed Chrome.

Hmmm... is this just a Chrome issue or a more general issue?  One of the main features of any sandbox is, of course, to isolate storage.

**Comment by strx5** [comment]
I have just tried to run the Portable version of the latest release on Windows 7 x64 and had to deal with crashes and errors. I think I've figured out what caused them so here's my Quick start guide for Sandboxie-Plus 0.7.1 x64 on Win 7 x64:

1. Run the installer: 64bit with 64bit version of  Windows;
2. Select "Extract all files to a directory for a portable use";
3. ! Temporaly rename the "32" subfolder in the root folder of Sandboxie;
4. Open the "Provisional.Windows.7.Drivers" archive and extract the "w7_install" cmd script and the SbieDrv-files into the root folder of Sandboxie;
5. Run the cmd script: it will replace the existing "SbieDrv.sys" driver with a copy of "SbieDrv.sys.5.48.5.x64.rc4";
If the script finds the "32" subfolder it will copy the 32-bit driver ("SbieDrv.sys.5.48.5.x32.rc4") _instead of the 64-bit one_ which will **crash** Sandboxie if you attempt to configure it, launch apps, etc
6. ! Rename the "32" subfolder once again (see p.12 below);
7. Run "SandMan.exe" (the sandbox manager); 
8. SandMan will launch a system service: Sandboxie Service (_the 64bit version_ of SbieSvc.exe in the root folder);
9. Additionally, SandMan will create two subkeys in the Windows Registry:
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\SbieDrv
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\SbieSvc
! SandMan will delete the subkeys on exit. If it fails to do so, remove them manually and **restart** Windows.
10. Press the "Cogs" button on the toolbar to access the global settings, on "The General Options Tab" enable "Add 'Run Sandboxed' to the explorer context menu";
11. On the "Advanced Option" tab choose the file system root folder for the Sandbox. All the changes made by apps running sandboxed will end up in the specified folder, e.g. 
if you download a file and save it to "D:\Downloads" the file will go to ..\'YourSandboxFolder'\'TheSandboxName'(DefaultSandbox, for instance)\Drive\D\Downloads;
12. ! When starting a 32-bit app in a sandbox, Sandboxie will search for _the 32-bit version_ of "SbieSvc.exe" in the "32" subfolder. 
If there is no such folder, you will run into multiple "**SBIE2101**" errors.

**Comment by Gitoffthelawn** [comment]
@strx5 Wow. That's a great step-by-step. Thank you.

That's for "more-or-less portable mode" use, correct?  (The *more-or-less* being that entries get written to the Windows registry, but then hopefully get deleted.)

**Comment by eseb63** [comment]
pay attention, maybe the cmd file has been modified since strx5 wrote his tutorial, but step 3 is now wrong for the cmd file i found in the windows 7 drivers archive (for x64 install) : if you rename the "32" folder, the 32 bits driver will be copied instead of the 64 bits driver.

second point : if necessary , do update the SbieVer variable in the script to the right version of the windows7 drivers (it was outdated for mine)

third point : it may take a while before the  driver starts when you run SandMan.exe and you may encounter a crash if you attempt to modify the global settings before that. if that, be patient and retry...

**Comment by STaRDoGG** [comment]
Can't start the driver in the latest version (see screenshot). This is the error I get:

![image](https://user-images.githubusercontent.com/1524526/122673261-35ed2b80-d195-11eb-888a-4cb6582dd4e6.png)

Both .ini's are in the root folder. I manually installed the service and driver via the maintenance menu, then clicked on start driver.

**Comment by isaak654** [comment]
@STaRDoGG 
I suggest you to uninstall the driver via the Maintenance menu > Advanced > Uninstall Driver and reopening Sandman.exe the next time without admin rights (you need to avoid right click -> run as administrator). At this point, you should be able to install the driver again & start it from the Maintenance menu.

Let me know if it works for you.

**Comment by NikitaB5** [comment]
I'm not sure who else has this problem, but I am trying to open a app with sand boxie but it doesn't open. I shows it is running, but the app doesn't actually open. Can someone help me out please?

**Comment by isaak654** [comment]
> 
> I'm not sure who else has this problem, but I am trying to open a app with sand boxie but it doesn't open. I shows it is running, but the app doesn't actually open. Can someone help me out please?

Could you open a new issue with detailed information from [here](https://github.com/sandboxie-plus/Sandboxie/issues/new?assignees=&labels=&template=bug_report.md&title=)? Just replace the information in the preset text with yours.

**Comment by Gitoffthelawn** [comment]
Are there any updated instructions for "installing" the portable version?

**Comment by isaak654** [comment]
There is a portable option that allows to extract the files from the Plus Installer itself, just run it.

**Comment by Gitoffthelawn** [comment]
@isaak654 Wow, nice and simple now.  Thank you!  Just run the extractor (installer) and then run `sandman.exe`.... that's it?  If so, much appreciated how easy it is now.

BTW, anything special to get rid of the portable version already "installed" (Sandboxie-Plus portable (SbiePlus64-v0.3.5))?  Just delete it's folder?

**Comment by isaak654** [comment]
Before deleting the portable folder:
* Make sure to stop/delete SbieSvc and SbieDrv via the Maintenance menu on Sandman or using the command lines
* Make sure to save these files on a safe storage, whether you keep them there or not: Sandboxie.ini, Sandboxie-Plus.ini
* Make sure to not have other personal files there

**Comment by Gitoffthelawn** [comment]
@isaak654 Thanks so much.  For the portable version, the SbieSvc and SbieDrv will only be running if Sandman was run since the last reboot, correct?

**Comment by isaak654** [comment]
Sandboxie Service is set to run automatically at system startup, even if you didn't open Sandman.

**Comment by Gitoffthelawn** [comment]
@isaak654 Hmmm... Is the service somehow hidden from the Windows list of services?  I'm not seeing it listed in the list of Windows services at all.

**Comment by isaak654** [comment]
One of my portable installs:
![SbieSvc_services.msc](https://user-images.githubusercontent.com/12372772/149661765-c9d2db6e-600b-4b19-8eb2-32e4f142f77f.png)

**Comment by Gitoffthelawn** [comment]
@isaak654 Thank you.  Not sure why I'm not seeing it listed.  Since you're seeing it there, is it also listed in the system's registry?

**Comment by isaak654** [comment]
I think you already received your answer here: https://github.com/sandboxie-plus/Sandboxie/issues/62#issuecomment-754591627

Sandman allows to stop/delete SbieSvc and SbieDrv at your convenience, probably you just forgot it.

**Comment by Gitoffthelawn** [comment]
@isaak654 Thank you again for all your help.
